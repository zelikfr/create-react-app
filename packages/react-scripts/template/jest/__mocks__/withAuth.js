import React, { Component, createContext } from 'react'

const withAuth = () => (ComposedComponent) => {
  class WithAuth extends Component {
    render () {
      return <ComposedComponent {...this.props} />
    }
  }
  return WithAuth
}
export const AuthContext = createContext()

export default withAuth
