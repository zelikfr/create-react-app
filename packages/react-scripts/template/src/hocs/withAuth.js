import React, { Component, createContext } from 'react'
import { Redirect } from 'react-router-dom'
import PropTypes from 'prop-types'
import jwtDecode from 'jwt-decode'
import { v4 } from 'uuid'
import axios from 'axios'
import config from 'config'

export const AuthContext = createContext()

const withAuth = (opts) => (ComposedComponent) => {
  const options = {
    redirectUri: '/',
    accessRestriction: 'any',
    ...opts,
  }
  class WithAuth extends Component {
    render () {
      const props = { ...this.props }
      return (
        <AuthContext.Consumer>
          {({ isAuthenticated, login, renewRefreshToken, logout }) => {
            if ((isAuthenticated && options.accessRestriction === 'notLogged') ||
              (!isAuthenticated && options.accessRestriction === 'logged')) {
              return <Redirect to={options.redirectUri} />
            }
            return <ComposedComponent {...props} login={login} logout={logout} renewRefreshToken={renewRefreshToken} />
          }}
        </AuthContext.Consumer>
      )
    }
  }

  return WithAuth
}

class AuthProviderComponent extends Component {
  constructor (props) {
    super(props)

    const refreshToken = window.localStorage.getItem('refreshToken')
    const expirationDate = window.localStorage.getItem('expirationDate')
    const userId = window.localStorage.getItem('userId')

    this.request = {}
    this.loginRequest = {}

    this.state = {
      isAuthenticated: !!refreshToken,
      refreshToken,
      expirationDate,
      userId,
    }
  }
  componentDidMount () {
    if (this.state.refreshToken) {
      this.renewRefreshToken()
    }
  }
  componentWillUnmount () {
    this._isMounted = false
    if (this.request.token) {
      this.request.cancel()
    }
    if (this.loginRequest.token) {
      this.loginRequest.cancel()
    }
  }
  authenticateUser (refreshToken) {
    window.localStorage.setItem('refreshToken', refreshToken)
    const { exp: expirationDate, global_user_id: userId } = jwtDecode(refreshToken)
    window.localStorage.setItem('refreshTokenExpirationDate', expirationDate)
    window.localStorage.setItem('userId', userId)
    this.setState({ isAuthenticated: true, expirationDate, userId })
  }
  logout () {
    window.localStorage.removeItem('refreshToken')
    window.localStorage.removeItem('refreshTokenExpirationDate')
    window.localStorage.removeItem('userId')
    this.setState({
      isAuthenticated: false,
      refreshToken: null,
      expirationDate: null,
      userId: null,
    })
  }
  async login (email, password) {
    if (this.loginRequest.token) {
      this.loginRequest.cancel()
    }
    this.loginRequest = axios.CancelToken.source()
    try {
      const { data: res } = await axios.request(
        {
          url: config.authUrl + '/api.companyusermasterdata.default/command',
          cancelToken: this.loginRequest.token,
          method: 'POST',
          data: {
            '$type': 'TraceOne.Api.CompanyUserMasterData.Default.Messages.Commands.AuthenticateUserCommand, TraceOne.Api.CompanyUserMasterData.Default.Messages',
            'messageId': v4(),
            email,
            password,
          },
        })

      if (res.succeeded) {
        this.authenticateUser(res.refreshToken)
        return true
      }
    } catch (err) {
      return err.error
    }
  }
  shouldRefreshToken (expirationDateToken) {
    return (expirationDateToken - Date.now() / 1000 < 7 * 60)
  }
  async renewRefreshToken () {
    if (this.shouldRefreshToken(this.state.expirationDate)) {
      if (this.request.token) {
        this.request.cancel()
      }
      this.request = axios.CancelToken.source()
      try {
        const { data: res } = await axios.request({
          url: config.authUrl + '/api.companyusermasterdata.default/command',
          method: 'POST',
          cancelToken: this.request.token,
          data: {
            '$type': 'TraceOne.Api.CompanyUserMasterData.Default.Messages.Commands.RenewRefreshTokenCommand, TraceOne.Api.CompanyUserMasterData.Default.Messages',
            'messageId': v4(),
            'RefreshToken': `${this.state.refreshToken}`,
            'UserId': `${this.state.userId}`,
          },
        })

        if (res.succeeded) {
          this.authenticateUser(res.refreshToken)
          return true
        }
      } catch (err) {
        if (!axios.isCancel(err)) {
          this.logout()
        }
      }
    }
    return false
  }
  render () {
    const { isAuthenticated } = this.state
    return (
      <AuthContext.Provider
        value={{
          isAuthenticated,
          login: (email, password) => this.login(email, password),
          renewRefreshToken: () => this.renewRefreshToken(),
          logout: () => this.logout(),
        }}
      >
        {this.props.children}
      </AuthContext.Provider>
    )
  }
}

AuthProviderComponent.propTypes = {
  children: PropTypes.any,
}

export const AuthProvider = (AuthProviderComponent)

export default withAuth
