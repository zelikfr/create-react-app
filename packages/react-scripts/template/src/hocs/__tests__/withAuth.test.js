import React from 'react'
import { mount, shallow } from 'enzyme'
import withAuth, { AuthContext, AuthProvider } from '../withAuth.js'
import axios from 'axios'

jest.mock('react-router-dom', () => ({
  Redirect: () => 'redirect',
}))

const Component = () => (
  <div></div>
)

describe('withAuth', () => {
  it('should create a component wrapped with withAuth without options', () => {
    // When I render the withAuth element
    const HocComponent = withAuth()(Component)
    const component = mount(
      <AuthContext.Provider
        value={{
          isAuthenticated: true,
          login: () => jest.fn(),
          logout: () => jest.fn(),
          renewRefreshToken: () => jest.fn(),
        }}
      >
        <HocComponent />
      </AuthContext.Provider>
    )
    // Then I should have the component to match the snapshot.
    expect(component).toMatchSnapshot()
  })
  it('should create a component wrapped with withAuth with options', () => {
    // When I render the withAuth element
    const HocComponent = withAuth({ redirectUri: '/', accessRestriction: 'logged' })(Component)
    const component = mount(
      <AuthContext.Provider
        value={{
          isAuthenticated: true,
          login: () => jest.fn(),
          logout: () => jest.fn(),
          renewRefreshToken: () => jest.fn(),
        }}
      >
        <HocComponent />
      </AuthContext.Provider>
    )
    // Then I should have the component to match the snapshot.
    expect(component).toMatchSnapshot()
  })
  it('should redirect when not authenticated and request authenticated', () => {
    // When I render the withAuth element and do a call
    const HocComponent = withAuth({ redirectUri: '/', accessRestriction: 'logged' })(Component)
    const component = mount(
      <AuthContext.Provider
        value={{
          isAuthenticated: false,
          login: () => jest.fn(),
          logout: () => jest.fn(),
          renewRefreshToken: () => jest.fn(),
        }}
      >
        <HocComponent />
      </AuthContext.Provider>,
      { selector: 'WithAuth' }
    )
    expect(component).toMatchSnapshot()
  })
  it('should redirect when authenticated and request not authenticated', () => {
    // When I render the withAuth element and do a call
    const HocComponent = withAuth({ redirectUri: '/', accessRestriction: 'notLogged' })(Component)
    const component = mount(
      <AuthContext.Provider
        value={{
          isAuthenticated: true,
          login: () => jest.fn(),
          logout: () => jest.fn(),
          renewRefreshToken: () => jest.fn(),
        }}
      >
        <HocComponent />
      </AuthContext.Provider>,
      { selector: 'WithAuth' }
    )
    expect(component).toMatchSnapshot()
  })
  it('should display when not authenticated and request not authenticated', () => {
    // When I render the withAuth element and do a call
    const HocComponent = withAuth({ redirectUri: '/', accessRestriction: 'notLogged' })(Component)
    const component = shallow(
      <AuthContext.Provider
        value={{
          isAuthenticated: false,
          login: () => jest.fn(),
          logout: () => jest.fn(),
          renewRefreshToken: () => jest.fn(),
        }}
      >
        <HocComponent />
      </AuthContext.Provider>,
      { selector: 'WithAuth' }
    )
    expect(component).toMatchSnapshot()
  })
})
describe('AuthProvider', () => {
  it('should create an AuthProvider wrapped component', () => {
    // When I render the withAuth element
    const component = mount(
      <AuthProvider />
    )
    // Then I should have the component to match the snapshot.
    expect(component).toMatchSnapshot()
  })
  it('should unmount an AuthProvider wrapped component without refreshToken', () => {
    // When I render the withAuth element
    const component = mount(
      <AuthProvider />
    )
    component.unmount()
    // Then I should have the component to match the snapshot.
    expect(component).toMatchSnapshot()
  })
  it('should unmount an AuthProvider wrapped component with refreshToken', () => {
    // When I render the withAuth element
    const component = mount(
      <AuthProvider />
    )
    component.instance().request.token = 'fakeToken'
    const spyCancel = jest.fn()
    component.instance().loginRequest.token = 'fakeToken'
    const spyLoginCancel = jest.fn()
    component.instance().request.cancel = spyCancel
    component.instance().loginRequest.cancel = spyLoginCancel
    component.unmount()
    // Then I should have the component to match the snapshot.
    expect(spyCancel).toHaveBeenCalled()
    expect(spyLoginCancel).toHaveBeenCalled()
  })
  it('should create an AuthProvider wrapped component with refreshToken', () => {
    // When I render the withAuth element
    jest.spyOn(window.Storage.prototype, 'getItem').mockImplementation((key) => {
      switch (key) {
        case 'refreshToken':
          return 'abctoken'
        case 'expirationDate':
          return '21/06/2019'
        case 'userId':
          return '123'
        default:
      }
    })
    const spyRenewRefresh = jest.spyOn(AuthProvider.prototype, 'renewRefreshToken')
    mount(
      <AuthProvider />
    )
    // Then I should have the component to match the snapshot.
    expect(spyRenewRefresh).toHaveBeenCalled()
    spyRenewRefresh.mockRestore()
  })
  it('should authenticate user correctly', () => {
    // When I render the withAuth element
    const spyStorage = jest.spyOn(window.Storage.prototype, 'setItem')
    const component = mount(
      <AuthProvider />
    )

    component.instance().authenticateUser('abc')
    // Then I should have the component to match the snapshot.
    expect(component.state('isAuthenticated')).toBe(true)
    expect(spyStorage).toHaveBeenCalled()
  })
  it('should logout user correctly', () => {
    // When I render the withAuth element
    const spyStorage = jest.spyOn(window.Storage.prototype, 'removeItem')
    const component = mount(
      <AuthProvider />
    )

    component.instance().logout()
    // Then I should have the component to match the snapshot.
    expect(component.state('isAuthenticated')).toBe(false)
    expect(spyStorage).toHaveBeenCalled()
  })
  it('should login user correctly', done => {
    // When I render the withAuth element
    const promise = Promise.resolve({ data: { succeeded: true, refreshToken: 'abcToken' } })
    const spyRequest = jest.spyOn(axios, 'request').mockImplementation(() => promise)
    const component = mount(
      <AuthProvider />
    )
    const spyAuth = jest.spyOn(component.instance(), 'authenticateUser').mockImplementation(() => jest.fn())
    component.instance().login('mail', 'pass')
    // Then I should have the component to match the snapshot.
    expect(spyRequest).toHaveBeenCalled()
    promise.then(() => {
      expect(spyAuth).toHaveBeenCalled()
      spyAuth.mockRestore()
      done()
    })
  })
  it('should login user badly due to the api', done => {
    // When I render the withAuth element
    const errorValue = { error: 'error' }
    const promise = Promise.reject(errorValue)
    const spyRequest = jest.spyOn(axios, 'request').mockImplementation(() => promise)
    const component = mount(
      <AuthProvider />
    )
    const spyAuth = jest.spyOn(component.instance(), 'authenticateUser').mockImplementation(() => jest.fn())
    component.instance().login('mail', 'pass')
    // Then I should have the component to match the snapshot.
    expect(spyRequest).toHaveBeenCalled()
    promise.then(() => {}).catch(() => {
      expect(spyAuth).not.toHaveBeenCalled()
      spyAuth.mockRestore()
      done()
    })
  })
  it('should login user correctly after cancel the current request', done => {
    // When I render the withAuth element
    const promise = Promise.resolve({ data: { succeeded: true, refreshToken: 'abcToken' } })
    const spyRequest = jest.spyOn(axios, 'request').mockImplementation(() => promise)
    const component = mount(
      <AuthProvider />
    )

    component.instance().loginRequest.token = 'fakeToken'
    const spyLoginCancel = jest.fn()
    component.instance().loginRequest.cancel = spyLoginCancel
    const spyAuth = jest.spyOn(component.instance(), 'authenticateUser').mockImplementation(() => jest.fn())
    component.instance().login('mail', 'pass')
    // Then I should have the component to match the snapshot.
    expect(spyLoginCancel).toHaveBeenCalled()
    expect(spyRequest).toHaveBeenCalled()
    promise.then(() => {
      expect(spyAuth).toHaveBeenCalled()
      spyAuth.mockRestore()
      done()
    })
  })
  it('should not login user with bad credentials', done => {
    // When I render the withAuth element
    const promise = Promise.resolve({ data: { succeeded: false } })
    const spyRequest = jest.spyOn(axios, 'request').mockImplementation(() => promise)
    const component = mount(
      <AuthProvider />
    )
    const spyAuth = jest.spyOn(component.instance(), 'authenticateUser').mockImplementation(() => jest.fn())
    component.instance().login('mail', 'pass')
    // Then I should have the component to match the snapshot.
    expect(spyRequest).toHaveBeenCalled()
    promise.then(() => {
      expect(spyAuth).not.toHaveBeenCalled()
      spyAuth.mockRestore()
      done()
    })
  })
  it('should return false when given date is after the current date', () => {
    // When I render the withAuth element
    const component = mount(
      <AuthProvider />
    )
    const currentDate = 1562346880373
    jest.spyOn(Date, 'now').mockImplementation(() => currentDate)
    const tokenDate = 1564346880
    // Then I should have the component to match the snapshot.
    expect(component.instance().shouldRefreshToken(tokenDate)).toBe(false)
  })
  it('should return true when given date is after the current date', () => {
    // When I render the withAuth element
    const component = mount(
      <AuthProvider />
    )
    const currentDate = 1562346880373
    jest.spyOn(Date, 'now').mockImplementation(() => currentDate)
    const tokenDate = 1561346880
    // Then I should have the component to match the snapshot.
    expect(component.instance().shouldRefreshToken(tokenDate)).toBe(true)
  })
  it('should return false and not renew the refreshToken', async () => {
    // When I render the withAuth element
    const component = mount(
      <AuthProvider />
    )
    const spyShould = jest.spyOn(component.instance(), 'shouldRefreshToken').mockImplementation(() => false)
    // Then I should have the component to match the snapshot.
    expect(await component.instance().renewRefreshToken()).toBe(false)
    spyShould.mockRestore()
  })
  it('should renew the refreshToken', async () => {
    // When I render the withAuth element
    const component = mount(
      <AuthProvider />
    )
    const spyShould = jest.spyOn(component.instance(), 'shouldRefreshToken').mockImplementation(() => true)
    // Then I should have the component to match the snapshot.

    const promise = Promise.resolve({ data: { succeeded: true, refreshToken: 'abcToken' } })
    const spyRequest = jest.spyOn(axios, 'request').mockImplementation(() => promise)
    expect(await component.instance().renewRefreshToken()).toBe(true)
    expect(spyRequest).toHaveBeenCalled()
    spyShould.mockRestore()
  })
  it('should renew the refreshToken and cancel current calls', async () => {
    // When I render the withAuth element
    const component = mount(
      <AuthProvider />
    )
    component.instance().request.token = 'fakeToken'
    const spyCancel = jest.fn()
    component.instance().request.cancel = spyCancel
    const spyShould = jest.spyOn(component.instance(), 'shouldRefreshToken').mockImplementation(() => true)
    // Then I should have the component to match the snapshot.

    const promise = Promise.resolve({ data: { succeeded: true, refreshToken: 'abcToken' } })
    const spyRequest = jest.spyOn(axios, 'request').mockImplementation(() => promise)
    expect(await component.instance().renewRefreshToken()).toBe(true)
    expect(spyCancel).toHaveBeenCalled()
    expect(spyRequest).toHaveBeenCalled()
    spyShould.mockRestore()
  })
  it('should not renew the refreshToken when not succeeded', async () => {
    // When I render the withAuth element
    const component = mount(
      <AuthProvider />
    )
    const spyShould = jest.spyOn(component.instance(), 'shouldRefreshToken').mockImplementation(() => true)
    // Then I should have the component to match the snapshot.

    const promise = Promise.resolve({ data: { succeeded: false } })
    const spyRequest = jest.spyOn(axios, 'request').mockImplementation(() => promise)
    expect(await component.instance().renewRefreshToken()).toBe(false)
    expect(spyRequest).toHaveBeenCalled()
    spyShould.mockRestore()
  })
  it('should not renew the refreshToken when API error and logout the user', async done => {
    // When I render the withAuth element
    const component = mount(
      <AuthProvider />
    )
    const spyLogout = jest.spyOn(component.instance(), 'logout')
    const spyShould = jest.spyOn(component.instance(), 'shouldRefreshToken').mockImplementation(() => true)
    // Then I should have the component to match the snapshot.

    const errorValue = { error: 'error' }
    const promise = Promise.reject(errorValue)
    const spyRequest = jest.spyOn(axios, 'request').mockImplementation(() => promise)
    expect(await component.instance().renewRefreshToken()).toBe(false)
    promise.then(() => { }).catch(() => {
      expect(spyRequest).toHaveBeenCalled()
      expect(spyLogout).toHaveBeenCalled()
      spyShould.mockRestore()
      done()
    })
  })
  it('should not renew the refreshToken when API error but not logout the user', async done => {
    // When I render the withAuth element
    const component = mount(
      <AuthProvider />
    )
    const spyLogout = jest.spyOn(component.instance(), 'logout')
    const spyShould = jest.spyOn(component.instance(), 'shouldRefreshToken').mockImplementation(() => true)
    // Then I should have the component to match the snapshot.
    jest.spyOn(axios, 'isCancel').mockImplementation(() => true)
    const errorValue = { error: 'error' }
    const promise = Promise.reject(errorValue)
    const spyRequest = jest.spyOn(axios, 'request').mockImplementation(() => promise)
    expect(await component.instance().renewRefreshToken()).toBe(false)
    promise.then(() => { }).catch(() => {
      expect(spyRequest).toHaveBeenCalled()
      expect(spyLogout).not.toHaveBeenCalled()
      spyShould.mockRestore()
      done()
    })
  })
  it('should call actions from provider', () => {
    // When I render the withAuth element
    const component = mount(
      <AuthProvider>
        <AuthContext.Consumer>
          {({ login, renewRefreshToken, logout }) => {
            return (
              <>
                <button className="login" onClick={() => login()} />
                <button className="logout" onClick={() => logout()} />
                <button className="renewRefreshToken" onClick={() => renewRefreshToken()} />
              </>
            )
          }}
        </AuthContext.Consumer>
      </AuthProvider>
    )
    const spyLogout = jest.spyOn(component.instance(), 'logout').mockImplementation(() => jest.fn())
    const spyLogin = jest.spyOn(component.instance(), 'login').mockImplementation(() => jest.fn())
    const spyRenew = jest.spyOn(component.instance(), 'renewRefreshToken').mockImplementation(() => jest.fn())
    // Then I should have the component to match the snapshot.
    component.find('.login').simulate('click')
    expect(spyLogin).toHaveBeenCalled()
    component.find('.logout').simulate('click')
    expect(spyLogout).toHaveBeenCalled()
    component.find('.renewRefreshToken').simulate('click')
    expect(spyRenew).toHaveBeenCalled()
  })
})
