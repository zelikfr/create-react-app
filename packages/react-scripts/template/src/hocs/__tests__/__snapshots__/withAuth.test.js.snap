// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthProvider should create an AuthProvider wrapped component 1`] = `<AuthProviderComponent />`;

exports[`AuthProvider should unmount an AuthProvider wrapped component without refreshToken 1`] = `null`;

exports[`withAuth should create a component wrapped with withAuth with options 1`] = `
<WithAuth>
  <Component
    login={[Function]}
    logout={[Function]}
    renewRefreshToken={[Function]}
  >
    <div />
  </Component>
</WithAuth>
`;

exports[`withAuth should create a component wrapped with withAuth without options 1`] = `
<WithAuth>
  <Component
    login={[Function]}
    logout={[Function]}
    renewRefreshToken={[Function]}
  >
    <div />
  </Component>
</WithAuth>
`;

exports[`withAuth should display when not authenticated and request not authenticated 1`] = `<WithAuth />`;

exports[`withAuth should redirect when authenticated and request not authenticated 1`] = `
<WithAuth>
  <Redirect
    to="/"
  >
    redirect
  </Redirect>
</WithAuth>
`;

exports[`withAuth should redirect when not authenticated and request authenticated 1`] = `
<WithAuth>
  <Redirect
    to="/"
  >
    redirect
  </Redirect>
</WithAuth>
`;
